src/Microsoft.OData.Core/Batch/ODataBatchWriter.cs

```diff
@@ -185,13 +186,13 @@ public void WriteEndBatch()
 
         /// <summary>Asynchronously ends a batch; can only be called after WriteStartBatch has been called and if no other active change set or operation exist.</summary>
         /// <returns>A task instance that represents the asynchronous write operation.</returns>
-        public Task WriteEndBatchAsync()
+        public async Task WriteEndBatchAsync()
         {
             this.VerifyCanWriteEndBatch(false);
-            return TaskUtils.GetTaskForSynchronousOperation(this.WriteEndBatchImplementation)
-
+            await this.WriteEndBatchImplementationAsync()
                 // Note that we intentionally go through the public API so that if the Flush fails the writer moves to the Error state.
-                .FollowOnSuccessWithTask(task => this.FlushAsync());
+                .FollowOnSuccessWithTask(task => this.FlushAsync())

```

Comment:
Out of curiosity, do you know why `ContinueWith` is not used here instead?

Reply:
Or just using `await` on the second statement, i.e.

```c#
await this.WriteEndBatchImplementationAsync();
await this.FlushAsync();
```

What else is `FollowOnSuccessWithTask` doing behind the scenes?

Reply:
You can think of `FollowOnSuccessWithTask` as a "syntactic" sugar for `ContinueWith`, although it's a little more than that. We actually use `ContinueWith` in the method. The method is designed to propagate failures if the antecedent task fails. If `WriteEndBatchImplementationAsync` throws an `ODataException`, we propagate that to the caller - not an `AggregateException` like `ContinueWith` would do.
The delegate passed to `FollowOnSuccessWithTask` will not be executed if the antecedent task fails. Specifically in this case, we do not wish to execute `FlushAsync` if `WriteEndBatchImplementationAsync` fails.
The method is defined in [`TaskUtils`](https://github.com/OData/odata.net/blob/master/src/Microsoft.OData.Core/TaskUtils.cs) class just in case you'd like to study the implementation further.